build/source/engine/read_attrb.f90

```diff
@@ -128,69 +127,61 @@ USE get_ixname_module,only:get_ixAttr,get_ixType  ! access function to find inde
+ allocate(gru_id(nGRU),stat=err)
  if(err/=0)then; err=20; message=trim(message)//'problem allocating space for GRU indices'; return; endif
 
- ! allocate space for hru-gru mapping vectors
- allocate(hru_id(nHRU),hru_ix(nHRU),hru2gru_id(nHRU),hru2gru_ixGRU(nHRU),hru2gru_ixHRU(nHRU), stat=err)
+ ! get hru_dim dimension length (ie., the number of hrus in entire domain)
+ err = nf90_inq_dimid(ncid, "hru", varid);                if(err/=0)then; message=trim(message)//'problem finding hru dimension/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_inquire_dimension(ncid, varid, len = maxHRU); if(err/=0)then; message=trim(message)//'problem reading hru dimension/'//trim(nf90_strerror(err)); return; endif  
+ ! allocate space for hru-gru mapping vectors 
+ allocate(hru_id(maxHRU),hru2gru_id(maxHRU), stat=err)
  if(err/=0)then; err=20; message=trim(message)//'problem allocating space for gru-hru mapping vectors'; return; endif
 
  ! **********************************************************************************************
  ! (2) read mapping vectors and populate mapping structures
  ! **********************************************************************************************
 
- ! read gru_id from netcdf file
- err = nf90_inq_varid(ncid, "gruId", varid); if(err/=0)then; message=trim(message)//'problem finding gruId'; return; endif
- err = nf90_get_var(ncid,varid,gru_id);      if(err/=0)then; message=trim(message)//'problem reading gruId'; return; endif
 
  ! read hruIndex from netcdf file
- err = nf90_inq_varid(ncid, "hruId", varid); if(err/=0)then; message=trim(message)//'problem finding hruId'; return; endif
- err = nf90_get_var(ncid,varid,hru_id);      if(err/=0)then; message=trim(message)//'problem reading hruId'; return; endif
+ err = nf90_inq_varid(ncid, "hruId", varid); if(err/=nf90_noerr)then; message=trim(message)//'problem finding hruId/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_get_var(ncid,varid,hru_id);      if(err/=nf90_noerr)then; message=trim(message)//'problem reading hruId/'//trim(nf90_strerror(err)); return; endif
 
  ! read hru2gru_id from netcdf file
- err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=0)then; message=trim(message)//'problem finding hru2gruId'; return; endif
- err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=0)then; message=trim(message)//'problem reading hru2gruId'; return; endif
-
- ! define the HRU and GRU indices
- gru_ix = arth(1,1,nGRU)
- hru_ix = arth(1,1,nHRU)
-
- ! create local mapping vectors
- do iGRU=1,nGRU
-  ! check
-  if(count(hru2gru_id == gru_id(iGRU))/=gru_struc(iGRU)%hruCount)then
-   write(message,'(a,1x,i0)') trim(message)//'unexpected number of HRUs for GRU', iGRU
-   err=20; return
-  endif
-  ! define indices
-  where(hru2gru_id == gru_id(iGRU))
-   hru2gru_ixGRU = gru_ix(iGRU)                       ! index of the GRU, for which a global HRU is within
-   hru2gru_ixHRU = arth(1,1,gru_struc(iGRU)%hruCount) ! indices of the local HRUs for each GRU
```

Comment:
I think this won't work when there are multiple GRUs, which leads to `size(hru2gru_ixHRU) /= gru_struc(iGRU)%hruCount`.

A special case of single GRU will warrant this line so the error has not been detected.

I would suggest adding a multiple-GRU simulation to the test cases in the future.